Split auditors

budang opened the pull request:
changes made per our discussion last week; still working on throttling, but otherwise everything should be good

Commit:
paste-event auditors

Commit:
removed the 'e' (event) param from binding function in paste scripts

Commit:
fixed some stuff in paste scripts; wrote separate scripts for keypress handlers and click handlers

Commit:
add separated scripts for scrolling auditors

Commit:
jQuery does not have separate event handlers for vertical and horizontal scrolling; added if statements to scroll auditors to only update their respective fields if their current_position != previous_position; this prevents a vertical scrolling handler from triggering during a horizontal scroll event and vice versa

Commit:
mouse movement auditors

Commit:
within_typing_delay auditor

Commit:
fixed math error in calculating time passed before comparison with delay

Commit:
*_specific.js all now return their respective arrays instead of null if the arrays are empty

Commit:
remove e.stopPropagation(), unneeded after all; clicks_specific now returns an array of Objects that describe a DOM element rather than an array of DOMs themselves

Commit:
removed e.stopPropagation(); changed binding event from keydown to keyup (in case someone holds down a key, we only want to count it once); removed 'originalEvent' in pastes_specific.js, unneeded

Commit:
removed e.stopPropagation(), unneeded

Commit:
removed unneeded fields

Commit:
merged horizontal and vertical scroll auditors into combined auditor and added setup functions; this currently works nicely, but I will look into throttling these as well

budang commented:
working on throttling for mouse_movement auditors and scrolling auditors

CuriousG102 commented:
Ok. I'll take a look tomorrow. Any trouble with the Python end or just not there yet? By the way, I fixed an efficiency issue with save_processed_data_to_model, so don't feel the need to override it only for efficiency's sake yet. I'll end up pulling the rug out from underneath you. I am making other changes to how our models work because I have so far overlooked tying step and auditor submissions to a specific instance of a user completing a task. Eventually I'll have to figure out how validation will tie in in a way that we don't have two completely separate code bases.

CuriousG102 commented:
If you need any help with debouncing let me know. I've had to do it before and it's not bad :p.

budang commented:
Haven't gotten to python part yet :neutral_face: once I finish up debouncing I will get to it lol. I'll let you know if I need help, I've had partial success but maybe I just need a little more time before getting it right completely

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/clicks_specific.js:
```diff
@@ -0,0 +1,33 @@
+    log_click_content: function (e) {
+        var dom = {
+            'type'  :   e.target.nodeName.toLowerCase(),
+            'id'    :   e.target.id != '' && e.target.id != undefined 
```
When you make corresponding Python/Django models, make sure to see the appropriate fields nullable to avoid problems.

budang replied:
:+1: 

Do you think empty string would be easier to work with?

CuriousG102 replied:
I'd keep it null. Making fields nullable is as simple as setting kwarg `null=True` in a field constructor. And it will be faster if we decide to aggregate on nulls in any way.

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/clicks_specific.js:
```diff
@@ -0,0 +1,33 @@
+        this.clicks_specifics.push(dom);
+    },
+    submit_callable: function () {
+        return {
```
This method of return can be altered in a way that prevents you from having to override base class behavior when you get to the Python stage. Change is simple. Instead of returning an Object, which will be converted to a dictionary, return an array of Objects with key/values corresponding to your eventual data model.

SO all you really need to do is 
`return this.clicks_specifics`

See _EventAndSubmissionModel.save_processed_data_to_model for more detail if desired

budang replied:
:+1:

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/keypresses_specific.js:
```diff
@@ -0,0 +1,24 @@
+var AuditorKeypressesSpecific = {
+    keypresses_specific: [],
+    log_keypress_content: function (e) {
+        this.keypresses_specific.push(String.fromCharCode(e.which));
```
Is this what you were talking about as a concern? LGTM.

CuriousG102 replied:
On second thought:
What does this look like when I hit backspace or enter? Shift? Tab?

budang replied:
Forgot to change it before making PR.

Respectively, `\b`, `\r`, `\u000e or \u000f`, `\t`.

CuriousG102 replied:
Cool. Then I stick by LGTM.

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/keypresses_specific.js:
```diff
@@ -0,0 +1,24 @@
+    },
+    submit_callable: function () {
+        return {
+            'content':  this.keypresses_specific
```
This is going to need some kind of modification vis a vis conversion to Python objects unless you want to write custom code A LOT. Which seems like an unnecessary way to muddy abstractions because we can do what I've seen so far without breaking abstractions.

So I think this one is less straightforward, particularly because I'm not sure how much resolution we want on this data or how backspaces/enters/etc will be handled. I see two extremes right now, though I'm not sure either one is the right idea.
One-
Model:
    content = models.TextField() # TextField of all keypresses

and being one model per HIT

Or -
Model:
    keypress = models.IntegerField() # e.which
    timestamp = models.IntegerField() # when user pressed

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/mouse_movement_specific.js:
```diff
@@ -0,0 +1,26 @@
+            'content':  this.mouse_movement_specific
+        };
+    }
+};
```
Same basic thing here. This javascript can be mapped a lot closer to what models will look like and that will save you and I a lot of code/pain (especially if we end up having to change parts of how Models work and we can just modify parents without worrying about downstream code.

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/clicks_total.js:
```diff
@@ -0,0 +1,24 @@
+var AuditorClicksTotal = {
```
LGTM

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/keypresses_total.js:
```diff
@@ -0,0 +1,24 @@
+var AuditorKeypressesTotal = {
```
LGTM

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/mouse_movement_total.js:
```diff
@@ -0,0 +1,25 @@
+var AuditorMouseMovementTotal = {
```
LGTM

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/pastes_specific.js:
```diff
@@ -0,0 +1,25 @@
+    },
+    submit_callable: function () {
+        return {
+            'content':  this.pastes_specific
```
Same thing vis a vis Python here.

Suggestion:
Model:
paste_content = models.TextField()
timestamp = models.DateTimeField() # idk if we actually want this field

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/pastes_total.js:
```diff
@@ -0,0 +1,24 @@
+var AuditorPastesTotal = {
```
LGTM

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/scrolled_pixels_specifics.js:
```diff
@@ -0,0 +1,48 @@
+
+        this.scrolled_pixels_specific.push(
+            {
+                'horizontal'    : horizontal,
```
Is it useful to preserve data on sequencing/time of these movements or no? Same for all others.

budang replied:
So you're saying add a timestamp field? And we should include a timestamp field for all the other auditors as well?

CuriousG102 replied:
What are your thoughts? We could do other stuff to preserve sequencing, like giving objects numbers 1, 2, 3, etc. Or we could abandon sequence of scrolls. But if we want to preserve, and we need an integer to do that, why not make it a timestamp?

budang replied:
I like timestamps. Would this apply to auditors for clicking, keypressing, etc as well or did you mean just for scrolling auditors? I'm thinking it would probably make more sense if probably everything had a time stamp

(Off topic, but we could present a chronological list of all the user's actions chronologically; maybe something we could look into at the very end? or not lol)

CuriousG102 replied:
Chronology would be interesting but it overlaps with export I mentioned in my to do list so let's hold off on that. When you get into the python models for Auditors, it might be a good idea to make an abstract model to inherit from (`TimeStampedAuditorDataModel`?) with the timestamp field. Then we can add behavior around that common field and guarantee consistency across data models using Timestamp.

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/scrolled_pixels_specifics.js:
```diff
@@ -0,0 +1,48 @@
+    },
+    submit_callable: function () {
+        return {
+            'content':  this.scrolled_pixels_specific
```
`return this.scrolled_pixels_specific`

budang replied:
:+1:

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/scrolled_pixels_specifics.js:
```diff
@@ -0,0 +1,48 @@
+auditor_scrolled_pixels_specific.setup();
+
+$(document).ready(function() {
+    $(window).scroll(auditor_scrolled_pixels_specific
```
Needs debounce to avoid race conditions and prodigious amounts of data
https://developer.mozilla.org/en-US/docs/Web/Events/scroll

budang replied:
Working on all debouncing stuff atm.

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/scrolled_pixels_total.js:
```diff
@@ -0,0 +1,49 @@
+auditor_scrolled_pixels_total.setup();
+
+$(document).ready(function() {
+    $(window).scroll(auditor_scrolled_pixels_total
```
Should have exact same debounce as above to avoid problems.

budang replied:
Working on all debouncing stuff atm.

Commit:
returning arrays instead of objects now for these auditors

Commit:
returning arrays instead of objects where possible

Commit:
returning arrays instead of objects where possible

Commit:
initial phase of debouncing; removed waiting for document ready since these events are based on interactions with the window instead of the document

Commit:
initial phase of debouncing; removed waiting for document ready since these events are based on interactions with the window instead of the document

Commit:
phase 2 of debounce; made debounce js into a separate static file and added it to survey-default-template html

Commit:
Merge branch 'master' of https://github.com/CuriousG102/turkey into split-auditors

Commit:
debouncing stuff

Commit:
added debouncing js script tag to html

Commit:
fixed address of debounce js file in the html

budang commented:
debouncing finished

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/mouse_movement_specific.js:
```diff
@@ -0,0 +1,26 @@
+        }
+    },
+    submit_callable: function () {
+        return this.mouse_movement_specific;
```
This needs to be an array of javascript objects, not an array of arrays. The Django code defaults to converting each object of the array to a Python dictionary, whose keys and values are then matched against field names and values of a django model.

Anything else requires overriding save code from the parent.

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/pastes_specific.js:
```diff
@@ -0,0 +1,23 @@
+        this.pastes_specific.push(pasted_data);
+    },
+    submit_callable: function () {
+        return this.pastes_specific;
```
Same thing here. Objects.

CuriousG102 commented on web/turkey/survey/static/survey/js/auditors/scrolled_pixels_specific.js:
```diff
@@ -0,0 +1,48 @@
+        previous_position_v = current_position_v;
+
+        this.scrolled_pixels_specific.push(
+            {
```
Good example of what to be doing. This is what I'm talking about.

CuriousG102 commented on web/turkey/survey/static/survey/js/globals/jquery.ba-throttle-debounce.min.js:
```diff
@@ -0,0 +1,9 @@
+ * Dual licensed under the MIT and GPL licenses.
+ * http://benalman.com/about/license/
+ */
+(function(b,c){var $=b.jQuery||b.Cowboy||(b.Cowboy={}),a;$.throttle=a=function(e,f,j,i){var h,d=0;if(typeof f!=="boolean"){i=j;j=f;f=c}function g(){var o=this,m=+new Date()-d,n=arguments;function l(){d=+new Date();j.apply(o,n)}function k(){h=c}if(i&&!h){l()}h&&clearTimeout(h);if(i===c&&m>e){l()}else{if(f!==true){h=setTimeout(i?k:l,i===c?e-m:e)}}}if($.guid){g.guid=j.guid=j.guid||$.guid++}return g} ...
```
Lol

CuriousG102 replied:
I was thinking about committing source so if Ben Alman's site goes down this isn't a complete black box. But I saw how big the source was when I went there and it got me to thinking: Why aren't we compressing our assets?

How do you feel about django-pipeline?
http://django-pipeline.readthedocs.org/en/latest/index.html

budang replied:
Just a side comment, but this project of Ben's seems to be a little neglected; last updated in 2010 with 6 issues and 5 PRs, oldest one dating from 2011 (nothing serious though).

django-pipeline seems fine, I don't see a reason not to use it lol. If you've used it before then I'm extra cool with it

CuriousG102 replied:
Ok, so sounds like it might be a good idea to have non-minified source. I'm not particularly worried. Based on experience, it wouldn't take too much effort to write debounce code of our own.

CuriousG102 commented:
I found two spots where we're not returning objects which can be converted by the code in the Auditor class (which your models will be inheriting from). As far as I can tell there's no reason they can't be written in a way that will allow them to be processed by code in Auditor class. Aside from that, LGTM.

Commit:
mouse_movement_specific now returns array of objects instead of array of arrays (an entry is now represented as an object with an 'X' and 'Y' field)

Commit:
pastes_specific now returns array of objects (entry now represented by an object with a 'data' field)

Commit:
styling changes to keep coding style consistent

budang commented:
Fixed

CuriousG102 committed:
Update mouse_movement_specific.js

Make x and y lowercase in order to match Python naming convention when model is made
